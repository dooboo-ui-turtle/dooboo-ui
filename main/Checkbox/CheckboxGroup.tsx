
import React, { FC, useCallback, useEffect, useRef, useState } from 'react';
import Checkbox from './Checkbox';
import styled from 'styled-components/native';

interface customStyle {
  labelSize?: number;
  labelColor?: string;
  boxSize?: number;
  boxColor?: string;
}

interface onChangeEvent {
  checked: boolean;
  label: string;
}

export type CheckboxValueType = string | number | boolean;

export interface CheckboxOptionType {
  label: string | number;
  value: CheckboxValueType;
  disabled?: boolean;
  onChange?: (e: onChangeEvent) => void;
  customStyle?: customStyle;
}

export interface CheckboxGroupState {
  value: CheckboxValueType[];
  registeredValues: CheckboxValueType[];
}

export interface CheckboxGroupContext {
  registerValue?:(value: CheckboxValueType) => void;
  toggleOption?: (option: CheckboxOptionType) => void;
  value?: any;
  disabled?: boolean;
}

interface CheckboxGroupProps {
  defaultValues?: Array<CheckboxValueType>;
  values?: Array<CheckboxValueType>;
  options?: Array<CheckboxOptionType | string>;
  disabled?: boolean;
  onChange?: (checkedValue: Array<CheckboxValueType>) => void;
  commonCustomStyle?: customStyle;
}

export const GroupCheckboxContext = React.createContext<CheckboxGroupContext | null>(null);

const CheckboxGroup: FC<CheckboxGroupProps> = (props) => {
  const isMounted = useRef(false);
  const [values, setValues] = useState<CheckboxValueType[]>(props.values || props.defaultValues || []);
  const [registeredValues, setRegisteredValues] = useState<CheckboxValueType[]>([]);

  useEffect(() => {
    // for only componenUpdate not in mount
    if (isMounted.current) {
      setValues(props.values || []);
    } else {
      isMounted.current = true;
    }
  }, [props.values]);

  const getOptions = useCallback((): Array<CheckboxOptionType> => {
    const { options } = props;

    return (options as Array<CheckboxOptionType>).map((option) => {
      if (typeof option === 'string') {
        return {
          label: option,
          value: option,
        } as CheckboxOptionType;
      }
      return option;
    });
  }, [props.options]);

  const registerValue = useCallback((value : string): void => {
    setRegisteredValues((prevRegisteredValues) => {
      return [...prevRegisteredValues, value];
    });
  }, [setRegisteredValues]);

  const toggleOption = (option: CheckboxOptionType): void => {
    const optionIndex = values.indexOf(option.value);
    const _values = [...values];
    if (optionIndex === -1) {
      _values.push(option.value);
    } else {
      _values.splice(optionIndex, 1);
    }
    if (!('values' in props)) {
      setValues(_values);
    }

    const { onChange } = props;
    if (onChange) {
      const options = getOptions();
      const filterdValues = _values
        .filter((val) => registeredValues.indexOf(val) !== -1)
        .sort((a, b) => {
          const indexA = options.findIndex((opt) => opt.value === a);
          const indexB = options.findIndex((opt) => opt.value === b);
          return indexA - indexB;
        });

      onChange(filterdValues);
    }
  };

  const context = {
    toggleOption,
    registerValue,
    values,
    disabled: props.disabled,
  };

  return (
    <Container>
      <GroupCheckboxContext.Provider value={context}>
        {getOptions().map((option) => (
          <Checkbox
            key={option.value.toString()}
            label={option.label}
            disabled={'disabled' in option ? option.disabled : props.disabled}
            value={option.value}
            checked={values.indexOf(option.value) !== -1}
            onChange={option.onChange}
            customStyle={ 'customStyle' in option ? option.customStyle : props.commonCustomStyle}
          >
            {option.label}
          </Checkbox>
        ))}
      </GroupCheckboxContext.Provider>
    </Container>

  );
};

const Container = styled.View`
 flex-direction: row;
 flex-wrap: wrap;
`;

export default CheckboxGroup;
