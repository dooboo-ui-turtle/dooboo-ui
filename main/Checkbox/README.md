# Checkbox

> Simple Checkbox for react-native.  
> Refer : <https://ant.design/components/checkbox/>

## Preview

| Checkbox | CheckboxGroup | CheckAll |   
|---------|-------------|------------|
| ![Checkbox](https://user-images.githubusercontent.com/45163013/92544089-fdc0d000-f287-11ea-9b46-1843566faad9.png) | ![CheckboxGroup](https://user-images.githubusercontent.com/45163013/92544095-00bbc080-f288-11ea-8d1d-cf95ab21bd22.png) | ![CheckAll](https://user-images.githubusercontent.com/45163013/92544100-01eced80-f288-11ea-8a29-557670e06c1d.png) |   

## Installation

```sh
yarn add dooboo-ui
```

## Props

### Common

```ts
interface customStyle {
  labelSize?: number;
  labelColor?: string;
  boxSize?: number;
  boxColor?: string;
}

interface onChangeEvent {
  checked: boolean;
  label: string | number;
}
```

### Checkbox

```ts

interface CheckboxProps {
  id?: string;
  label: string | number;
  value?: string | number | boolean;
  checked?: boolean
  defaultChecked?: boolean;
  indeterminate?: boolean;
  disabled?: boolean
  onChange?: (event: onChangeEvent) => void;
  customStyle?: customStyle;
}
```

### CheckboxGroup
```ts

type CheckboxValueType = string | number | boolean;

interface CheckboxOptionType {
  label: string | number;
  value: CheckboxValueType;
  disabled?: boolean;
  onChange?: (e: onChangeEvent) => void;
  customStyle?: customStyle;
}

interface CheckboxGroupState {
  value: CheckboxValueType[];
  registeredValues: CheckboxValueType[];
}

interface CheckboxGroupProps {
  defaultValues?: Array<CheckboxValueType>;
  values?: Array<CheckboxValueType>;
  options?: Array<CheckboxOptionType | string>;
  disabled?: boolean;
  onChange?: (checkedValue: Array<CheckboxValueType>) => void;
  commonCustomStyle?: customStyle;
}
```
## API

### Checkbox

| Property       |      Required      | Type                              | Default |
| -------------- | :----------------: | --------------------------------- | ------- |
| label          | :white_check_mark: | `string | number`                 |         |
| value          |                    | `string | number | boolean`       |         |
| checked        |                    | `boolean`                         | `fasle` |
| defaultChecked |                    | `boolean`                         | `false` |
| indeterminate  |                    | `boolean`                         | `false` |
| disabled       |                    | `boolean`                         | `false` |
| onChange       |                    | `(event: onChangeEvent) => void;` |         |
| customStyle    |                    | `customStyle type`                |         |


#### customStyle 
| Property   | Required | Type      | Default    |
| ---------- | :------: | --------- | ---------- |
| labelSize  |          | `number`  | `20`       |
| labelColor |          | `string`  | `#000000`, |
| boxSize    |          | `number`  | `20`       |
| boxColor   |          | `string`  | `#1890FF`  |
| labelLeft  |          | `boolean` | `#`        |

### CheckboxGroup
| Property          | Required | Type                                                | Default |
| ----------------- | :------: | --------------------------------------------------- | ------- |
| values            |          | `Array<CheckboxValueType>`                          |         |
| defaultValues     |          | `Array<CheckboxValueType>`                          |         |
| options           |          | `Array<CheckboxOptionType | string>`                |         |
| disabled          |          | `boolean`                                           |         |
| onChange          |          | `(event: onChangeEvent) => void;`                   |         |
| commonCustomStyle |          | `customStyle`                                       |         |


## Getting started

### Import

```tsx
import { Checkbox, CheckboxGroup } from 'dooboo-ui';
```

### Usage

#### Checkbox

```tsx
function DefaultCheckbox(): React.ReactElement {
  const customStyle = {
    labelLeft: true,
  };
  return (
    <ScrollContainer>

      <Container>
        <Title>default</Title>
        <Checkbox label="defaultChecked" defaultChecked ></Checkbox>

      </Container>

      <Container>
        <Title>disabled</Title>
        <Checkbox label="disabled" disabled />
      </Container>

      <Container>
        <Title>indeterminate</Title>
        <Checkbox label="indeterminate" indeterminate />
      </Container>

      <Container>
        <Title>labelLeft</Title>
        <Checkbox label="labelLeft" customStyle={customStyle} />
      </Container>

    </ScrollContainer>
  );
}
```

#### CheckboxGroup

```tsx
function DefaultCheckboxGroup(): React.ReactElement {
  const plainOptions = ['Apple', 'Pear', 'Orange'];
  const options = [
    { label: 'Apple', value: 'Apple' },
    { label: 'Pear', value: 'Pear' },
    { label: 'Orange', value: 'Orange' },
    { label: 'Mango', value: 'Mango' },
  ];
  const optionsWithDisabled = [
    { label: 'Apple', value: 'Apple' },
    { label: 'Pear', value: 'Pear' },
    { label: 'Orange', value: 'Orange', disabled: true },
  ];

  const onChange = (checkedValues): void => {
    console.log('checked = ', checkedValues);
  };
  return (
    <ScrollContainer>

      <Container>
        <Title>plainOption</Title>
        <CheckboxGroup options={plainOptions} onChange={onChange} />
      </Container>
      <Container>
        <Title>Default</Title>
        <CheckboxGroup options={options} onChange={onChange} />
      </Container>

      <Container>
        <Title>option with disabled</Title>
        <CheckboxGroup options={optionsWithDisabled} onChange={onChange} />
      </Container>

      <Container>
        <Title>with disabled props</Title>
        <CheckboxGroup options={options} disabled onChange={onChange} />
      </Container>

    </ScrollContainer>
  );
}
```
