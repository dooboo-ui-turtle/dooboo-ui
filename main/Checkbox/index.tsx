import React, { FC, useCallback, useState } from 'react';
import {
  TouchableHighlight,
} from 'react-native';

import styled from 'styled-components/native';

interface MarkerContainerProps {
  boxSize?: number
  boxColor?: string
}

interface LabelProps {
  labelColor?: string
  labelSize?: number
}

interface MarkerProps {
  isChecked: boolean
}

const Container = styled.View`
 flex-direction: row;
 align-items: center;
`;

const MarkerContainer = styled.View<MarkerContainerProps>`
 padding: 3px;
 width: ${({ boxSize }):number => boxSize || 20}px;
 height: ${({ boxSize }):number => boxSize || 20}px;
 background-color: ${({ boxColor }): string => boxColor || '#cecece'};
`;

const Marker = styled.View<MarkerProps>`
 flex: 1;
 justify-content: center;
 align-items: center;
 background-color: ${({ isChecked }): string => isChecked ? 'transparent' : '#ffffff'};
`;

const MarkerImg = styled.Image`
 width: 85%;
 height: 85%;
 tint-color: #ffffff;
 resize-mode: contain;
`;

const Label = styled.Text<LabelProps>`
 font-size: ${({ labelSize }):number => labelSize || 20}px;
 padding-left: 10px;
 color: ${({ labelColor }): string => labelColor || '#000000'};
`;

interface Props {
  boxSize?: number
  boxColor?: string
  defaultChecked?: boolean
  disabled?: boolean
  onStateChange?(): void
  labelText?: string
  labelSize?: number
  labelColor?: string
}

const Checkbox: FC<Props> = ({
  boxSize = 20,
  boxColor = '#cecece',
  labelText = 'default',
  labelSize = 20,
  labelColor = '#000000',
  defaultChecked = false,
  onStateChange = ():void => {},
  /* TODO  */
  // disabled = false,
}) => {
  const [isChecked, setIstChecked] = useState<boolean>(defaultChecked);
  const onPress = useCallback(() => {
    setIstChecked((prevIsChecked) => !prevIsChecked);
    onStateChange();
  }, [onStateChange]);
  return (
    <TouchableHighlight
      onPress={onPress}
      underlayColor="transparent"
      style={{ marginVertical: 20 }}>
      <Container>
        <MarkerContainer
          boxSize={boxSize}
          boxColor={boxColor}
        >
          <Marker isChecked={isChecked}>
            {isChecked && <MarkerImg
              source={require('../__assets__/check_tick.png')}
            />}
          </Marker>
        </MarkerContainer>
        <Label labelSize={labelSize} labelColor={labelColor}>
          {labelText}
        </Label>
      </Container>
    </TouchableHighlight>
  );
};

export { Checkbox };
