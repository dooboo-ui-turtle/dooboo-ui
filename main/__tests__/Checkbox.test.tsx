import * as React from 'react';
import { Checkbox } from '../Checkbox';
import { TouchableHighlight } from 'react-native';
import renderer from 'react-test-renderer';

let props: unknown;

// Note: test renderer must be required after react-native.
const createTestProps = (obj: Record<string, unknown>): Record<string, unknown> => ({
  navigation: {
    navigate: jest.fn(),
  },
  ...obj,
});

const component = (props?): React.ReactElement => {
  return <Checkbox {...props}/>;
};

describe('[CheckBox]', () => {
  beforeEach(() => {
    props = createTestProps({});
  });

  it('render without crashing', () => {
    const rendered: renderer.ReactTestRendererJSON | null = renderer
      .create(component(props))
      .toJSON();
    expect(rendered).toMatchSnapshot();
  });

  describe('[CheckBox] Interaction', (): void => {
    it('should simulate onPress', (): void => {
      const handlePress = jest.fn();
      const rendered = renderer.create(
        component({
          label: 'CHECKBOX_LABEL',
          onChange: handlePress,
        }),
      );

      const checkboxClick = rendered.root.findByType(TouchableHighlight);
      renderer.act(() => {
        checkboxClick.props.onPress();
      });
      expect(handlePress).toHaveBeenCalled();
    });
    it('should simulate props', (): void => {
      const rendered = renderer.create(
        component({
          label: 'CHECKBOX_LABEL',
        }),
      );

      rendered.update(component({ disabled: true }));
      expect(rendered).toMatchSnapshot();
      expect(rendered).toBeTruthy();
    });
  });
});
