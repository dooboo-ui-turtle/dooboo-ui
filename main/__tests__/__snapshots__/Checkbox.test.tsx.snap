// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[CheckBox] [CheckBox] Interaction should simulate props 1`] = `
<
  disabled={true}
  onPress={[Function]}
  style={
    Object {
      "marginHorizontal": 20,
      "paddingBottom": 20,
    }
  }
  underlayColor="transparent"
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "center",
        },
      ]
    }
  >
    <View
      boxColor="#E0E0E0"
      style={
        Array [
          Object {
            "backgroundColor": "#E0E0E0",
            "height": 20,
            "paddingBottom": 1.5,
            "paddingLeft": 1.5,
            "paddingRight": 1.5,
            "paddingTop": 1.5,
            "width": 20,
          },
        ]
      }
    >
      <View
        isChecked={false}
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "#ffffff",
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
              "justifyContent": "center",
            },
          ]
        }
      />
    </View>
    <Text
      labelColor="#E0E0E0"
      style={
        Array [
          Object {
            "color": "#E0E0E0",
            "fontSize": 20,
            "paddingLeft": 10,
            "paddingRight": 0,
          },
        ]
      }
    />
  </View>
</>
`;

exports[`[CheckBox] render without crashing 1`] = `
<
  disabled={false}
  onPress={[Function]}
  style={
    Object {
      "marginHorizontal": 20,
      "paddingBottom": 20,
    }
  }
  underlayColor="transparent"
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "center",
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#1890FF",
            "height": 20,
            "paddingBottom": 1.5,
            "paddingLeft": 1.5,
            "paddingRight": 1.5,
            "paddingTop": 1.5,
            "width": 20,
          },
        ]
      }
    >
      <View
        isChecked={false}
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "#ffffff",
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
              "justifyContent": "center",
            },
          ]
        }
      />
    </View>
    <Text
      style={
        Array [
          Object {
            "color": "#000000",
            "fontSize": 20,
            "paddingLeft": 10,
            "paddingRight": 0,
          },
        ]
      }
    />
  </View>
</>
`;
